git/config.go

```diff
@@ -134,17 +135,33 @@ func (c *Configuration) UnsetLocalKey(key string) (string, error) {
 	return c.gitConfigWrite("--unset", key)
 }
 
-func (c *Configuration) Sources(optionalFilename string) ([]*ConfigurationSource, error) {
+func (c *Configuration) Sources(dir string, optionalFilename string) ([]*ConfigurationSource, error) {
 	gitconfig, err := c.Source()
 	if err != nil {
 		return nil, err
 	}
 
-	fileconfig, err := c.FileSource(optionalFilename)
-	if err != nil && !os.IsNotExist(err) {
+	bare, err := IsBare()
+	if err != nil {
 		return nil, err
 	}
 
+	// First try to read from the working directory and then the index if
+	// the file is missing from the working directory.
+	var fileconfig *ConfigurationSource
+	if !bare {
+		fileconfig, err = c.FileSource(filepath.Join(dir, optionalFilename))
+		if err != nil {
+			if !os.IsNotExist(err) {
+				return nil, err
+			}
+			fileconfig, _ = c.RevisionSource(fmt.Sprintf(":%s", optionalFilename))
```

Comment:
Just as a question, is it the case that there are not interesting errors to report from trying to run `git config ...` here?  In the `c.FileSource()` call above it looks like `c.gitConfig()` is called and any errors from it get returned as long as they aren't `os.IsNotExist()`, but similar returns from `c.RevisionSource()` look to be ignored here and below in the `HEAD` call.

Reply:
Ah, I see you anticipated my question in the PR description!
> It's intentional that we ignore errors from this code path, since Git doesn't distinguish well between older versions which don't support the `--blob` flag to `git config` (those before 1.8.4), cases where the revision is missing, and actual errors.